### ISSUE TITLE
Stats Board: Average price wrong calculated

### ISSUE THREAD

**Issue by 01park** [PrestaShop/PrestaShop#25547]
#### Describe the bug

I just noticed that the average sales price in our store makes no sense. Until, I figured out it is calculated based on the forecast / amount of days in the period.

#### Expected behavior

What I would expect and what I used to have in previous PS versions, is that average sales is based on the total sales over a period / the days, months or years. 

#### Steps to Reproduce

Check the sales in stats dash and see if the average amount is correct.

1. Go to 'Stats'
2. Click on ' stats board'
3. Scroll down to 'average sales '
4. See error ( = incorrect value )

**Screenshots**

#### Additional information

* PrestaShop version: 1.7.7.5
* PHP version: 7.3

**Comment by hibatallahAouadni** [comment]
Hello @01park 

Thanks for respecting the provided template :clap: 
But I really need steps to reproduce the issue :pray: I need to find exactly the culprit that triggered the problem :wink: 
Please check and feedback.

Thanks!

**Comment by 01park** [comment]
Go to stats > stats board and see average sales

![Screenshot 2021-08-09 at 14 07 44](https://user-images.githubusercontent.com/46492443/128697569-08cb6130-5742-4ce1-a6e1-930c57e112d1.png)


I covered our sales, but his is the location

**Comment by hibatallahAouadni** [comment]
Hello @01park 

I need an exemple to understand better the issue, for exemple: **the averge sales should be ... but it display ... and it's wrong cause if you calculate it the total ... and divide into ... the result should be ...** .
Please check and feedback.

Thanks!

**Comment by 01park** [comment]
Hi,

The average sales is calculated by the forecast / the selected time period. It should be, the total sales / selected time period.

What now is being calculated is the forecasted total sales average sales and it should be the average sales of the total sales

**Comment by khouloudbelguith** [comment]
Hi @01park,

Thanks for your feedback.
I made a fresh install PS1777
I checked:
- The Average cart value in the ` Payment distribution` section => ok
- The Average price in the `Category distribution` section => ok

- The Average of the `STATS DASHBOARD` section, I have strange calculus.
![image](https://user-images.githubusercontent.com/16067358/130218389-a9465128-5cb9-4b09-ac66-953a489ec7f3.png)

I don't know exactly if they are ok or not => this is your exact issue.
I will ping our developers to check the issue.

Ping @PrestaShop/prestashop-maintainers please could you check this issue.

Thanks!

**Comment by 01park** [comment]
Hi @khouloudbelguith 

From my view, the average is being calculated from the forecast revenue / the select amount of time  ( for example 30 day ). I calculated this and everytime it is spot 

In your example I see you selected yearly. Not sure how its calculated here. 

I remember stats dashboard recently being updated, from what I can remember I was correct back then.

**Comment by matks** [comment]
> Ping @PrestaShop/prestashop-maintainers please could you check this issue.

Hello ! Sorry but I dont see how developers can help you sort this out 😅 . The people who built this are long gone and I guess there is no documentation about what's expected here. Unless someone digs the ghosts of the past to try to find out what happened, we'll never know.

I dont think it's interesting to dig in the past. I think the Product team should analyze this module and simply state what it should output, no matter the current implementation. And we should implement the correct behavior we chose.

**Comment by 01park** [comment]
Hi @matks 


 I assume for whole prestashop it's important that data is showen correctly. Showing an average sale, in my opinion, either should be correct, or removed by prestahop if they believe its out dated. Why show information that is incorrect?
 

The issue is also in 1.7.8

**Comment by boo-code** [comment]
@01park's observation is exactly right, and it is an identity rather than a coincidence. `statsforecast.php` computes, for the elapsed part of the selected range:

```
interval  = full selected range, in days
interval2 = elapsed part of the range, in days      (ends at the current time of day)
prop30    = interval / interval2                    Forecast = total * prop30
interval_avg = interval2 (daily) | /7 | /30 | /365  Average  = total / interval_avg
```

For the daily view `interval_avg` **is** `interval2`, so `Average * interval == Forecast`, exactly. Measured, for every range I tried:

```
                            forecast    avg * selected days   identical
today only                  1999.98     1999.98               yes
this month (day 6 of 30)    5454.54     5454.54               yes
this year (day 249 of 365)  1468.81     1468.81               yes
```

That is why the average looks like it is derived from the forecast. It is not - both are built on the same `interval2` denominator, so they move together.

The defect underneath is that denominator. `interval2` ends at the current time of day, so it is almost always fractional, while the rows it summarises are whole days, weeks, months or years. Dividing a whole-period total by a fraction of a period overstates the average, and the error is worst on short ranges. With a total of 1000 over the elapsed rows:

```
selected range (daily)      rows   divisor   average shown   mean of the rows   overstated
today only                  1      0.500     2000.00         1000.00            +100.0%
yesterday + today           2      1.500      666.67          500.00             +33.3%
this month, on day 6        6      5.500      181.82          166.67              +9.1%
last month, fully elapsed   31    31.000       32.26           32.26              +0.0%
```

The average of a table showing one row is twice that row. It converges only once the range is entirely in the past, which is why a fully elapsed period looks correct and the current one never does.

The other granularities have the same shape, and yearly is the clearest: for 1 Jan to today, `interval_avg` is `248.5 / 365 = 0.68`, a divisor below 1, so the yearly **average comes out larger than the total**. Weekly gives 35.5 where 36 weeks are shown, monthly 8.28 where 9 months are shown.

A range whose start is in the future makes `interval2` negative (-24.5 days for a range starting next month), so the averages come out negative and `prop30` inverts.

Prepared against `dev`: count the calendar periods that actually elapsed, grouped with the same key the queries group by, and divide by that. The average row then equals the mean of the rows above it, which is what the table claims it is.

```
                  old divisor   new divisor
daily             248.500       249
weekly             35.500        36
monthly             8.283         9
yearly              0.681         1
future range      -24.500         1
```

The weekly key mirrors `MAKEDATE(YEAR(d), DAYOFYEAR(d)-WEEKDAY(d))` including the `CONCAT(YEAR(d),'-01-01*')` fallback the query uses when that offset lands before 1 January; verified against MySQL across 2025-12-29 to 2026-01-06, identical for all nine days. For the daily view the new divisor equals the number of rows the table pre-seeds, checked on six ranges including one spanning a DST switch.

Two things deliberately left out. Empty periods still do not produce a row for the weekly, monthly and yearly views, because only the daily view pre-seeds its table - so those averages are over the periods that have data. Making them render zero rows would be the follow-up, and it is a display change rather than an arithmetic one. And this does not touch what the forecast means, only the divisor the average shares with it.

On @matks' point that the product team should state what the module ought to output: that is still true for the wider question, but it is not needed for this one. The average row sits directly under the rows it averages, so its meaning is fixed by the table itself - no specification is required to say that the mean of a single row of 1000 is not 2000.
